qiskit/primitives/base_estimator.py

```diff
@@ -227,12 +259,104 @@ def __call__(
+            QiskitError: For mismatch of length of Sequence.
         """
+
+        # Support ndarray
+        if isinstance(parameter_values, np.ndarray):
+            parameter_values = parameter_values.tolist()
+
+        # Allow lift T to [T].
+        if circuits is not None and not isinstance(circuits, (np.ndarray, Sequence)):
+            circuits = [circuits]
+        if observables is not None and not isinstance(observables, (np.ndarray, Sequence)):
+            observables = [observables]
+        if parameter_values is not None and not isinstance(
+            parameter_values[0], (np.ndarray, Sequence)
+        ):
+            parameter_values = cast("Sequence[float]", parameter_values)
+            parameter_values = [parameter_values]
+
+        # Allow broadcasting
+        if (
+            circuits is None
+            and len(self._circuits) == 1
+            and observables is None
+            and len(self._observables) == 1
+            and parameter_values is not None
+        ):
+            circuits = [0] * len(parameter_values)
+            observables = [0] * len(parameter_values)
+
+        # Allow optional
+        if circuits is None:
+            circuits = list(range(len(self._circuits)))
+        if observables is None:
+            observables = list(range(len(self._observables)))
+        if parameter_values is None:
+            parameter_values = [[]] * len(circuits)
+
+        # Allow objects
+        try:
+            circuits = [
+                next(_finditer(circuit.name, self._circuit_names))
+                if not isinstance(circuit, (int, np.integer))
+                else circuit
+                for circuit in circuits
+            ]
+        except StopIteration as err:
+            raise QiskitError(
+                "The circuits passed when calling estimator is not one of the circuits used to "
+                "initialize the session."
+            ) from err
+        try:
+            observables = [
+                next(_finditer(id(observable), self._observable_ids))
+                if not isinstance(observable, (int, np.integer))
+                else observable
+                for observable in observables
+            ]
+        except StopIteration as err:
+            raise QiskitError(
+                "The observables passed when calling estimator is not one of the circuits used to "
+                "initialize the session."
+            ) from err
+
+        # Validation
+        if len(circuits) != len(observables):
+            raise QiskitError(
+                f"The number of circuits ({len(circuits)}) does not match "
+                f"the number of observables ({len(observables)})."
+            )
+        if len(circuits) != len(parameter_values):
+            raise QiskitError(
+                f"The number of circuits ({len(circuits)}) does not match "
+                f"the number of parameter value sets ({len(parameter_values)})."
+            )
+
```

Comment:
```suggestion

        for circ_i, obs_i in zip(circuits, observables):
            circuit_num_qubits = self.circuits[circ_i].num_qubits
            observable_num_qubits = self.observables[obs_i].num_qubits
            if circuit_num_qubits != observable_num_qubits:
                raise QiskitError(
                    f"The number of qubits of the {circ_i}-th circuit ({circuit_num_qubits}) does "
                    f"not match the number of qubits of the {obs_i}-th observable "
                    f"({observable_num_qubits})."
                )

```
This is also in the internal implementation. Can this also be added to the base class?